File: internal/analyzers/burndown/metrics.go
package burndown

import (
	"time"

	"github.com/Sumatoshi-tech/codefang/internal/analyzers/analyze"
)

// Constants for burndown metrics calculations.
const (
	// Default tick size is 24 hours.
	defaultTickSizeHours = 24

	// Percent multiplier for calculations.
	percentMultiplier = 100

	// Modifier index offset for interaction calculations.
	modifierIndexOffset = 2

	// Self modifier ID.
	selfModifierID = -2
)

// ReportData is the parsed input data for burndown metrics computation.
type ReportData struct {
	GlobalHistory      DenseHistory
	FileHistories      map[string]DenseHistory
	FileOwnership      map[string]map[int]int
	PeopleHistories    []DenseHistory
	PeopleMatrix       DenseHistory
	ReversedPeopleDict []string
	TickSize           time.Duration
	Sampling           int
	Granularity        int
	ProjectName        string
	EndTime            time.Time
}

// ParseReportData extracts ReportData from an analyzer report.
func ParseReportData(report analyze.Report) (*ReportData, error) {
	data := &ReportData{}

	if gh, ok := report["GlobalHistory"].(DenseHistory); ok {
		data.GlobalHistory = gh
	}

	if fh, ok := report["FileHistories"].(map[string]DenseHistory); ok {
		data.FileHistories = fh
	}

	if fo, ok := report["FileOwnership"].(map[string]map[int]int); ok {
		data.FileOwnership = fo
	}

	if ph, ok := report["PeopleHistories"].([]DenseHistory); ok {
		data.PeopleHistories = ph
	}

	if pm, ok := report["PeopleMatrix"].(DenseHistory); ok {
		data.PeopleMatrix = pm
	}

	if rpd, ok := report["ReversedPeopleDict"].([]string); ok {
		data.ReversedPeopleDict = rpd
	}

	if ts, ok := report["TickSize"].(time.Duration); ok {
		data.TickSize = ts
	} else {
		data.TickSize = defaultTickSizeHours * time.Hour
	}

	if s, ok := report["Sampling"].(int); ok {
		data.Sampling = s
	}

	if g, ok := report["Granularity"].(int); ok {
		data.Granularity = g
	}

	if pn, ok := report["ProjectName"].(string); ok {
		data.ProjectName = pn
	}

	if et, ok := report["EndTime"].(time.Time); ok {
		data.EndTime = et
	}

	return data, nil
}

// SurvivalData contains code survival statistics for a time period.
type SurvivalData struct {
	SampleIndex   int     `json:"sample_index"   yaml:"sample_index"`
	TotalLines    int64   `json:"total_lines"    yaml:"total_lines"`
	SurvivalRate  float64 `json:"survival_rate"  yaml:"survival_rate"`
	BandBreakdown []int64 `json:"band_breakdown" yaml:"band_breakdown"`
}

// FileSurvivalData contains survival data for a single file.
type FileSurvivalData struct {
	Path         string      `json:"path"                 yaml:"path"`
	CurrentLines int64       `json:"current_lines"        yaml:"current_lines"`
	Ownership    map[int]int `json:"ownership"            yaml:"ownership"`
	TopOwnerID   int         `json:"top_owner_id"         yaml:"top_owner_id"`
	TopOwnerName string      `json:"top_owner_name"       yaml:"top_owner_name"`
	TopOwnerPct  float64     `json:"top_owner_percentage" yaml:"top_owner_percentage"`
}

// DeveloperSurvivalData contains survival data for a developer's code.
type DeveloperSurvivalData struct {
	ID           int     `json:"id"            yaml:"id"`
	Name         string  `json:"name"          yaml:"name"`
	CurrentLines int64   `json:"current_lines" yaml:"current_lines"`
	PeakLines    int64   `json:"peak_lines"    yaml:"peak_lines"`
	SurvivalRate float64 `json:"survival_rate" yaml:"survival_rate"`
}

// InteractionData contains developer interaction statistics.
type InteractionData struct {
	AuthorID      int    `json:"author_id"      yaml:"author_id"`
	AuthorName    string `json:"author_name"    yaml:"author_name"`
	ModifierID    int    `json:"modifier_id"    yaml:"modifier_id"`
	ModifierName  string `json:"modifier_name"  yaml:"modifier_name"`
	LinesModified int64  `json:"lines_modified" yaml:"lines_modified"`
	IsSelfModify  bool   `json:"is_self_modify" yaml:"is_self_modify"`
}

// AggregateData contains summary statistics.
type AggregateData struct {
	TotalCurrentLines   int64   `json:"total_current_lines"   yaml:"total_current_lines"`
	TotalPeakLines      int64   `json:"total_peak_lines"      yaml:"total_peak_lines"`
	OverallSurvivalRate float64 `json:"overall_survival_rate" yaml:"overall_survival_rate"`
	AnalysisPeriodDays  int     `json:"analysis_period_days"  yaml:"analysis_period_days"`
	NumBands            int     `json:"num_bands"             yaml:"num_bands"`
	NumSamples          int     `json:"num_samples"           yaml:"num_samples"`
	TrackedFiles        int     `json:"tracked_files"         yaml:"tracked_files"`
	TrackedDevelopers   int     `json:"tracked_developers"    yaml:"tracked_developers"`
}

func computeGlobalSurvival(input *ReportData) []SurvivalData {
	if len(input.GlobalHistory) == 0 {
		return nil
	}

	peakLines := findPeakLines(input.GlobalHistory)
	result := make([]SurvivalData, len(input.GlobalHistory))

	for i, sample := range input.GlobalHistory {
		result[i] = computeSurvivalSample(i, sample, peakLines)
	}

	return result
}

// findPeakLines computes the total lines ever written across all cohorts.
// Each band represents a cohort of code written during a time period.
// The peak of each band (its maximum value across all samples) is the number
// of lines originally written in that cohort. Summing these gives the total
// lines ever written, which is the correct denominator for survival rate.
func findPeakLines(history DenseHistory) int64 {
	if len(history) == 0 {
		return 0
	}

	numBands := len(history[0])
	bandPeaks := make([]int64, numBands)

	for _, sample := range history {
		for band := range min(len(sample), numBands) {
			if sample[band] > bandPeaks[band] {
				bandPeaks[band] = sample[band]
			}
		}
	}

	var total int64

	for _, peak := range bandPeaks {
		total += peak
	}

	return total
}

func sumPositiveValues(values []int64) int64 {
	var total int64

	for _, v := range values {
		if v > 0 {
			total += v
		}
	}

	return total
}

func computeSurvivalSample(index int, sample []int64, peakLines int64) SurvivalData {
	breakdown := make([]int64, len(sample))

	var total int64

	for j, v := range sample {
		if v > 0 {
			total += v
			breakdown[j] = v
		}
	}

	var survivalRate float64
	if peakLines > 0 {
		survivalRate = float64(total) / float64(peakLines)
	}

	return SurvivalData{
		SampleIndex:   index,
		TotalLines:    total,
		SurvivalRate:  survivalRate,
		BandBreakdown: breakdown,
	}
}

// FileSurvivalInput holds input for file survival computation.
type FileSurvivalInput struct {
	FileHistories      map[string]DenseHistory
	FileOwnership      map[string]map[int]int
	ReversedPeopleDict []string
}

// DeveloperSurvivalInput holds input for developer survival computation.
type DeveloperSurvivalInput struct {
	PeopleHistories    []DenseHistory
	ReversedPeopleDict []string
}

// InteractionInput holds input for interaction computation.
type InteractionInput struct {
	PeopleMatrix       DenseHistory
	ReversedPeopleDict []string
}

func computeFileSurvival(input FileSurvivalInput) []FileSurvivalData {
	result := make([]FileSurvivalData, 0, len(input.FileOwnership))

	for path, ownership := range input.FileOwnership {
		var (
			currentLines  int64
			topOwnerID    int
			topOwnerLines int
		)

		for devID, lines := range ownership {
			currentLines += int64(lines)
			if lines > topOwnerLines {
				topOwnerLines = lines
				topOwnerID = devID
			}
		}

		var topOwnerPct float64
		if currentLines > 0 {
			topOwnerPct = float64(topOwnerLines) / float64(currentLines) * percentMultiplier
		}

		topOwnerName := ""
		if topOwnerID >= 0 && topOwnerID < len(input.ReversedPeopleDict) {
			topOwnerName = input.ReversedPeopleDict[topOwnerID]
		}

		result = append(result, FileSurvivalData{
			Path:         path,
			CurrentLines: currentLines,
			Ownership:    ownership,
			TopOwnerID:   topOwnerID,
			TopOwnerName: topOwnerName,
			TopOwnerPct:  topOwnerPct,
		})
	}

	return result
}

func computeDeveloperSurvivalList(input DeveloperSurvivalInput) []DeveloperSurvivalData {
	result := make([]DeveloperSurvivalData, 0, len(input.PeopleHistories))

	for devID, history := range input.PeopleHistories {
		if len(history) == 0 {
			continue
		}

		data := computeDeveloperSurvival(devID, history, input.ReversedPeopleDict)
		result = append(result, data)
	}

	return result
}

func computeDeveloperSurvival(devID int, history DenseHistory, names []string) DeveloperSurvivalData {
	peakLines := findPeakLines(history)
	currentLines := sumPositiveValues(history[len(history)-1])

	var survivalRate float64
	if peakLines > 0 {
		survivalRate = float64(currentLines) / float64(peakLines)
	}

	name := ""
	if devID < len(names) {
		name = names[devID]
	}

	return DeveloperSurvivalData{
		ID:           devID,
		Name:         name,
		CurrentLines: currentLines,
		PeakLines:    peakLines,
		SurvivalRate: survivalRate,
	}
}

func computeInteraction(input InteractionInput) []InteractionData {
	if len(input.PeopleMatrix) == 0 {
		return nil
	}

	var result []InteractionData

	for authorID, row := range input.PeopleMatrix {
		if len(row) == 0 {
			continue
		}

		interactions := computeAuthorInteractions(authorID, row, input.ReversedPeopleDict)
		result = append(result, interactions...)
	}

	return result
}

func computeAuthorInteractions(authorID int, row []int64, names []string) []InteractionData {
	authorName := getName(authorID, names)

	var result []InteractionData

	for modifierIdx, lines := range row {
		if lines == 0 {
			continue
		}

		data := buildInteractionData(authorID, authorName, modifierIdx, lines, names)
		result = append(result, data)
	}

	return result
}

func buildInteractionData(authorID int, authorName string, modifierIdx int, lines int64, names []string) InteractionData {
	modifierID := modifierIdx - modifierIndexOffset
	isSelf := modifierID == selfModifierID

	modifierName := resolveModifierName(modifierID, authorName, isSelf, names)
	if isSelf {
		modifierID = authorID
	}

	return InteractionData{
		AuthorID:      authorID,
		AuthorName:    authorName,
		ModifierID:    modifierID,
		ModifierName:  modifierName,
		LinesModified: lines,
		IsSelfModify:  isSelf,
	}
}

func getName(id int, names []string) string {
	if id >= 0 && id < len(names) {
		return names[id]
	}

	return ""
}

func resolveModifierName(modifierID int, authorName string, isSelf bool, names []string) string {
	if isSelf {
		return authorName
	}

	return getName(modifierID, names)
}

func computeAggregate(input *ReportData) AggregateData {
	agg := AggregateData{
		TrackedFiles:      len(input.FileHistories),
		TrackedDevelopers: len(input.PeopleHistories),
	}

	if len(input.GlobalHistory) == 0 {
		return agg
	}

	agg.NumSamples = len(input.GlobalHistory)
	agg.NumBands = len(input.GlobalHistory[0])

	totalTicks := (agg.NumSamples - 1) * input.Sampling
	agg.AnalysisPeriodDays = int(time.Duration(totalTicks) * input.TickSize / (defaultTickSizeHours * time.Hour))

	agg.TotalPeakLines = findPeakLines(input.GlobalHistory)
	agg.TotalCurrentLines = sumPositiveValues(input.GlobalHistory[agg.NumSamples-1])

	if agg.TotalPeakLines > 0 {
		agg.OverallSurvivalRate = float64(agg.TotalCurrentLines) / float64(agg.TotalPeakLines)
	}

	return agg
}

// ComputedMetrics holds all computed metric results for the burndown analyzer.
type ComputedMetrics struct {
	Aggregate         AggregateData           `json:"aggregate"          yaml:"aggregate"`
	GlobalSurvival    []SurvivalData          `json:"global_survival"    yaml:"global_survival"`
	FileSurvival      []FileSurvivalData      `json:"file_survival"      yaml:"file_survival"`
	DeveloperSurvival []DeveloperSurvivalData `json:"developer_survival" yaml:"developer_survival"`
	Interaction       []InteractionData       `json:"interactions"       yaml:"interactions"`
}

// --- MetricsOutput Interface Implementation ---.

const analyzerNameBurndown = "burndown"

// AnalyzerName returns the analyzer identifier.
func (m *ComputedMetrics) AnalyzerName() string {
	return analyzerNameBurndown
}

// ToJSON returns the metrics in JSON-serializable format.
func (m *ComputedMetrics) ToJSON() any {
	return m
}

// ToYAML returns the metrics in YAML-serializable format.
func (m *ComputedMetrics) ToYAML() any {
	return m
}

// ComputeAllMetrics runs all burndown metrics and returns the results.
func ComputeAllMetrics(report analyze.Report) (*ComputedMetrics, error) {
	input, err := ParseReportData(report)
	if err != nil {
		return nil, err
	}

	globalSurvival := computeGlobalSurvival(input)

	fileSurvival := computeFileSurvival(FileSurvivalInput{
		FileHistories:      input.FileHistories,
		FileOwnership:      input.FileOwnership,
		ReversedPeopleDict: input.ReversedPeopleDict,
	})

	devSurvival := computeDeveloperSurvivalList(DeveloperSurvivalInput{
		PeopleHistories:    input.PeopleHistories,
		ReversedPeopleDict: input.ReversedPeopleDict,
	})

	interaction := computeInteraction(InteractionInput{
		PeopleMatrix:       input.PeopleMatrix,
		ReversedPeopleDict: input.ReversedPeopleDict,
	})

	aggregate := computeAggregate(input)

	return &ComputedMetrics{
		GlobalSurvival:    globalSurvival,
		FileSurvival:      fileSurvival,
		DeveloperSurvival: devSurvival,
		Interaction:       interaction,
		Aggregate:         aggregate,
	}, nil
}

File: rust/crates/cf-analyzer-burndown/src/metrics.rs
//! Burndown `ComputedMetrics` model and byte-identical serialization.
//!
//! Mirrors Go `internal/analyzers/burndown/metrics.go`: the per-analyzer
//! `ComputedMetrics` struct that `BaseHistoryAnalyzer.Serialize` marshals for the
//! `json` / `yaml` / `bin` machine formats. Field/key order follows the Go struct
//! declaration order (`aggregate`, `global_survival`, `file_survival`,
//! `developer_survival`, `interactions`), which the struct-origin [`GoMap`]
//! preserves.
//!
//! Go nil-slice asymmetry: `encoding/json` renders a nil slice as `null`, while
//! `gopkg.in/yaml.v3` renders both nil and empty slices as `[]`. The two builders
//! [`ComputedMetrics::to_go_value`] (JSON/bin) and
//! [`ComputedMetrics::to_go_value_yaml`] (YAML) differ only on this point.

use cf_gojson::{GoMap, GoValue};

/// Code-survival statistics for one sample (Go `SurvivalData`).
#[derive(Clone, Debug, Default, PartialEq)]
pub struct SurvivalData {
    /// `sample_index`.
    pub sample_index: i64,
    /// `total_lines`.
    pub total_lines: i64,
    /// `survival_rate`.
    pub survival_rate: f64,
    /// `band_breakdown`.
    pub band_breakdown: Vec<i64>,
}

impl SurvivalData {
    fn to_go_value(&self) -> GoValue {
        let mut m = GoMap::new_struct();
        m.push("sample_index", GoValue::Int(self.sample_index));
        m.push("total_lines", GoValue::Int(self.total_lines));
        m.push("survival_rate", GoValue::Float(self.survival_rate));
        m.push(
            "band_breakdown",
            GoValue::Array(self.band_breakdown.iter().map(|&v| GoValue::Int(v)).collect()),
        );
        GoValue::Map(m)
    }
}

/// Summary statistics (Go `AggregateData`).
#[derive(Clone, Debug, Default, PartialEq)]
pub struct AggregateData {
    /// `total_current_lines`.
    pub total_current_lines: i64,
    /// `total_peak_lines`.
    pub total_peak_lines: i64,
    /// `overall_survival_rate`.
    pub overall_survival_rate: f64,
    /// `analysis_period_days`.
    pub analysis_period_days: i64,
    /// `num_bands`.
    pub num_bands: i64,
    /// `num_samples`.
    pub num_samples: i64,
    /// `tracked_files`.
    pub tracked_files: i64,
    /// `tracked_developers`.
    pub tracked_developers: i64,
}

impl AggregateData {
    fn to_go_value(&self) -> GoValue {
        let mut m = GoMap::new_struct();
        m.push("total_current_lines", GoValue::Int(self.total_current_lines));
        m.push("total_peak_lines", GoValue::Int(self.total_peak_lines));
        m.push("overall_survival_rate", GoValue::Float(self.overall_survival_rate));
        m.push("analysis_period_days", GoValue::Int(self.analysis_period_days));
        m.push("num_bands", GoValue::Int(self.num_bands));
        m.push("num_samples", GoValue::Int(self.num_samples));
        m.push("tracked_files", GoValue::Int(self.tracked_files));
        m.push("tracked_developers", GoValue::Int(self.tracked_developers));
        GoValue::Map(m)
    }
}

/// All computed burndown metric results (Go `ComputedMetrics`).
///
/// `file_survival`, `developer_survival`, and `interactions` are modeled as
/// optional vectors; `None` reproduces a Go nil slice (`null` in JSON, `[]` in
/// YAML), `Some` an explicit (possibly empty) slice (`[]` in both).
#[derive(Clone, Debug, Default, PartialEq)]
pub struct ComputedMetrics {
    /// `aggregate`.
    pub aggregate: AggregateData,
    /// `global_survival`.
    pub global_survival: Vec<SurvivalData>,
    /// `file_survival`: `None` = Go nil slice.
    pub file_survival: Option<Vec<GoValue>>,
    /// `developer_survival`: `None` = Go nil slice.
    pub developer_survival: Option<Vec<GoValue>>,
    /// `interactions`: `None` = Go nil slice.
    pub interactions: Option<Vec<GoValue>>,
}

impl ComputedMetrics {
    fn build(&self, nil_as_empty: bool) -> GoValue {
        let nil = |opt: &Option<Vec<GoValue>>| -> GoValue {
            match opt {
                Some(v) => GoValue::Array(v.clone()),
                None => {
                    if nil_as_empty {
                        GoValue::Array(Vec::new())
                    } else {
                        GoValue::Null
                    }
                }
            }
        };

        let mut m = GoMap::new_struct();
        m.push("aggregate", self.aggregate.to_go_value());
        m.push(
            "global_survival",
            GoValue::Array(self.global_survival.iter().map(SurvivalData::to_go_value).collect()),
        );
        m.push("file_survival", nil(&self.file_survival));
        m.push("developer_survival", nil(&self.developer_survival));
        m.push("interactions", nil(&self.interactions));
        GoValue::Map(m)
    }

    /// Build the JSON/bin value tree (Go `encoding/json`: nil slice → `null`).
    #[must_use]
    pub fn to_go_value(&self) -> GoValue {
        self.build(false)
    }

    /// Build the YAML value tree (`gopkg.in/yaml.v3`: nil slice → `[]`).
    #[must_use]
    pub fn to_go_value_yaml(&self) -> GoValue {
        self.build(true)
    }
}

/// Sparse global history accumulated across the commit walk:
/// `curTick -> prevTick -> lineCountDelta`. Mirrors the Go `sparseHistory`
/// (`map[int]map[int]int64`) that the burndown aggregator builds by additively
/// merging every per-commit `CommitResult.GlobalDeltas`.
pub type SparseHistory = std::collections::BTreeMap<i64, std::collections::BTreeMap<i64, i64>>;

/// Dense burndown matrix (Go `DenseHistory = [][]int64`): rows are samples,
/// columns are cohort bands.
pub type DenseHistory = Vec<Vec<i64>>;

/// `groupSparseHistory` (Go `internal/analyzers/burndown/history_dense.go`):
/// densify a sparse tick history into a `samples x bands` matrix. `sampling` and
/// `granularity` are the band/sample sizes (both 30 by default); `last_tick` is
/// the maximum tick observed across the whole walk.
#[must_use]
pub fn group_sparse_history(history: &SparseHistory, sampling: i64, granularity: i64, last_tick: i64) -> DenseHistory {
    if history.is_empty() {
        return Vec::new();
    }

    // normalizeTicks: sorted tick keys (BTreeMap already sorted); append
    // last_tick if it exceeds the largest key (last_tick >= 0 here).
    let mut ticks: Vec<i64> = history.keys().copied().collect();
    let resolved_last = if let Some(&max_key) = ticks.last() {
        if max_key < last_tick {
            ticks.push(last_tick);
            last_tick
        } else {
            last_tick
        }
    } else {
        last_tick.max(0)
    };

    let samples = (resolved_last / sampling + 1) as usize;
    let bands = (resolved_last / granularity + 1) as usize;

    let mut result: DenseHistory = vec![vec![0i64; bands]; samples];

    // fillDenseHistory: carry forward state across empty sample rows, then add
    // each tick's per-band deltas into its sample row.
    let mut prevsi: usize = 0;
    for &tick in &ticks {
        let si = (tick / sampling) as usize;
        if si > prevsi {
            let state = result[prevsi].clone();
            for row in result.iter_mut().take(si + 1).skip(prevsi + 1) {
                row.copy_from_slice(&state);
            }
            prevsi = si;
        }
        if let Some(row) = history.get(&tick) {
            for (&t, &value) in row {
                let band = (t / granularity) as usize;
                if band < bands {
                    result[si][band] += value;
                }
            }
        }
    }

    result
}

/// `findPeakLines` (Go `metrics.go`): sum, over every band, of that band's
/// maximum value across all samples — the total lines ever written.
fn find_peak_lines(history: &DenseHistory) -> i64 {
    if history.is_empty() {
        return 0;
    }
    let num_bands = history[0].len();
    let mut band_peaks = vec![0i64; num_bands];
    for sample in history {
        for band in 0..sample.len().min(num_bands) {
            if sample[band] > band_peaks[band] {
                band_peaks[band] = sample[band];
            }
        }
    }
    band_peaks.iter().sum()
}

/// `sumPositiveValues` (Go `metrics.go`).
fn sum_positive(values: &[i64]) -> i64 {
    values.iter().filter(|&&v| v > 0).sum()
}

/// `computeSurvivalSample` (Go `metrics.go`).
fn compute_survival_sample(index: i64, sample: &[i64], peak_lines: i64) -> SurvivalData {
    let mut breakdown = vec![0i64; sample.len()];
    let mut total = 0i64;
    for (j, &v) in sample.iter().enumerate() {
        if v > 0 {
            total += v;
            breakdown[j] = v;
        }
    }
    let survival_rate = if peak_lines > 0 { total as f64 / peak_lines as f64 } else { 0.0 };
    SurvivalData { sample_index: index, total_lines: total, survival_rate, band_breakdown: breakdown }
}

/// `ComputeAllMetrics` (Go `metrics.go`) for the default report shape
/// (`PeopleNumber == 0`, `TrackFiles == false`): only `GlobalHistory`,
/// `Sampling`, `Granularity`, and `TickSize` feed the output, so
/// `file_survival` / `developer_survival` stay empty and `interactions` nil.
///
/// `tick_size_hours` is the configured tick size in hours (24 by default);
/// `analysis_period_days = (num_samples-1) * sampling * tick_size_hours / 24`,
/// matching Go's `time.Duration` arithmetic in `computeAggregate`.
#[must_use]
pub fn compute_global_metrics(global_dense: &DenseHistory, sampling: i64, tick_size_hours: i64) -> ComputedMetrics {
    // computeGlobalSurvival.
    let peak = find_peak_lines(global_dense);
    let global_survival: Vec<SurvivalData> = global_dense
        .iter()
        .enumerate()
        .map(|(i, sample)| compute_survival_sample(i as i64, sample, peak))
        .collect();

    // computeAggregate (TrackedFiles/TrackedDevelopers are 0: no per-file /
    // per-people histories in the default report).
    let mut aggregate = AggregateData::default();
    if !global_dense.is_empty() {
        let num_samples = global_dense.len() as i64;
        aggregate.num_samples = num_samples;
        aggregate.num_bands = global_dense[0].len() as i64;
        let total_ticks = (num_samples - 1) * sampling;
        // time.Duration(totalTicks) * tickSize / 24h, integer-truncated.
        aggregate.analysis_period_days = total_ticks * tick_size_hours / 24;
        aggregate.total_peak_lines = peak;
        aggregate.total_current_lines = sum_positive(&global_dense[num_samples as usize - 1]);
        if aggregate.total_peak_lines > 0 {
            aggregate.overall_survival_rate =
                aggregate.total_current_lines as f64 / aggregate.total_peak_lines as f64;
        }
    }

    ComputedMetrics {
        aggregate,
        global_survival,
        // computeFileSurvival over an empty FileOwnership map → empty slice
        // (make([]FileSurvivalData, 0, ...)); same for developer_survival.
        file_survival: Some(Vec::new()),
        developer_survival: Some(Vec::new()),
        // computeInteraction over an empty PeopleMatrix → nil slice.
        interactions: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn head_metrics(n: i64) -> ComputedMetrics {
        ComputedMetrics {
            aggregate: AggregateData {
                total_current_lines: n,
                total_peak_lines: n,
                overall_survival_rate: 1.0,
                analysis_period_days: 0,
                num_bands: 1,
                num_samples: 1,
                tracked_files: 0,
                tracked_developers: 0,
            },
            global_survival: vec![SurvivalData {
                sample_index: 0,
                total_lines: n,
                survival_rate: 1.0,
                band_breakdown: vec![n],
            }],
            file_survival: Some(Vec::new()),
            developer_survival: Some(Vec::new()),
            interactions: None,
        }
    }

    #[test]
    fn json_matches_golden_shape() {
        let bytes = cf_gojson::marshal(&head_metrics(112_539).to_go_value());
        let s = String::from_utf8(bytes).unwrap();
        assert_eq!(
            s,
            r#"{"aggregate":{"total_current_lines":112539,"total_peak_lines":112539,"overall_survival_rate":1,"analysis_period_days":0,"num_bands":1,"num_samples":1,"tracked_files":0,"tracked_developers":0},"global_survival":[{"sample_index":0,"total_lines":112539,"survival_rate":1,"band_breakdown":[112539]}],"file_survival":[],"developer_survival":[],"interactions":null}"#
        );
    }

    #[test]
    fn yaml_renders_nil_slice_as_empty() {
        let bytes = cf_goyaml::marshal(&head_metrics(112_539).to_go_value_yaml());
        let s = String::from_utf8(bytes).unwrap();
        assert!(s.contains("interactions: []"), "{s}");
        assert!(s.contains("file_survival: []"), "{s}");
        assert!(s.contains("    total_current_lines: 112539"), "{s}");
        assert!(s.contains("        - 112539"), "{s}");
    }
}
